benchmark/Bench.hs

```diff
@@ -15,27 +16,45 @@ main = defaultMain
       ]
     , bgroup "standalone"
-      [ bench "100"       $ whnf (run . execWriter @(Sum Int) . tellLoop) 100
-      , bench "1000"    $ whnf (run . execWriter @(Sum Int) . tellLoop) 1000
+      [ bench "100"   $ whnf (run . execWriter @(Sum Int) . tellLoop) 100
+      , bench "1000"  $ whnf (run . execWriter @(Sum Int) . tellLoop) 1000
       , bench "10000" $ whnf (run . execWriter @(Sum Int) . tellLoop) 10000
       ]
     ]
   ,
     bgroup "Strict StateC"
     [ bgroup "Cod"
-      [ bench "100"       $ whnf (run . runCod pure . execState @(Sum Int) 0 . runCod pure . modLoop) 100
-      , bench "1000"    $ whnf (run . runCod pure . execState @(Sum Int) 0 . runCod pure . modLoop) 1000
-      , bench "10000" $ whnf (run . runCod pure . execWriter @(Sum Int) . runCod pure . tellLoop) 10000
+      [ bench "100"   $ whnf (run . runCod pure . execState @(Sum Int) 0 . runCod pure . modLoop) 100
+      , bench "1000"  $ whnf (run . runCod pure . execState @(Sum Int) 0 . runCod pure . modLoop) 1000
+      , bench "10000" $ whnf (run . runCod pure . execState @(Sum Int) 0 . runCod pure . modLoop) 10000
       ]
     , bgroup "standalone"
-      [ bench "100"       $ whnf (run . execState @(Sum Int) 0 . modLoop) 100
-      , bench "1000"    $ whnf (run . execState @(Sum Int) 0 . modLoop) 1000
-      , bench "10000" $ whnf (run . execWriter @(Sum Int) . tellLoop) 10000
+      [ bench "100"   $ whnf (run . execState @(Sum Int) 0 . modLoop) 100
+      , bench "1000"  $ whnf (run . execState @(Sum Int) 0 . modLoop) 1000
+      , bench "10000" $ whnf (run . execState @(Sum Int) 0 . modLoop) 10000
+      ]
+    ]
+  ,
+    bgroup "InterpretC vs InterpretStateC vs StateC"
+    [ bgroup "InterpretC"
+      [ bench "100"   $ whnf (run . evalState @(Sum Int) 0 . runInterpret (\case { Get k -> get @(Sum Int) >>= k ; Put s k -> put s >> k }) . modLoop) 100
+      , bench "1000"  $ whnf (run . evalState @(Sum Int) 0 . runInterpret (\case { Get k -> get @(Sum Int) >>= k ; Put s k -> put s >> k }) . modLoop) 1000
+      , bench "10000" $ whnf (run . evalState @(Sum Int) 0 . runInterpret (\case { Get k -> get @(Sum Int) >>= k ; Put s k -> put s >> k }) . modLoop) 10000
+      ]
+    , bgroup "InterpretStateC"
+      [ bench "100"   $ whnf (run . runInterpretState (\ s -> \case { Get k -> runState @(Sum Int) s (k s) ; Put s k -> runState s k }) 0 . modLoop) 100
+      , bench "1000"  $ whnf (run . runInterpretState (\ s -> \case { Get k -> runState @(Sum Int) s (k s) ; Put s k -> runState s k }) 0 . modLoop) 1000
+      , bench "10000" $ whnf (run . runInterpretState (\ s -> \case { Get k -> runState @(Sum Int) s (k s) ; Put s k -> runState s k }) 0 . modLoop) 10000
+      ]
+    , bgroup "StateC"
+      [ bench "100"   $ whnf (run . evalState @(Sum Int) 0 . modLoop) 100
+      , bench "1000"  $ whnf (run . evalState @(Sum Int) 0 . modLoop) 1000
+      , bench "10000" $ whnf (run . evalState @(Sum Int) 0 . modLoop) 10000
```

Comment:
In terms of elapsed time, `StateC` < `InterpretStateC` < `InterpretC`. Lesson: use `runInterpret` to experiment, use a new `Carrier` for efficiency.

Reply:
Interestingly enough, `InterpretC` appears to be faster for me, with `-O2` on:

```
InterpretC/1000:  65.67 μs
InterpretStateC/1000: 133.9 μs
StateC/1.938 μs
```

Reply:
Huh, I hadn’t encountered that. I wonder if the passthru is getting optimized (partly) away?